File: android/src/main/kotlin/school/attractor/attractor_geo/upload/PointJson.kt
package school.attractor.attractor_geo.upload

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import org.json.JSONArray
import org.json.JSONObject
import school.attractor.attractor_geo.db.PointRow

/** Builds the upload body. The wire format is fixed by the design spec. */
object PointJson {
    private val iso8601: SimpleDateFormat
        get() = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

    fun encode(points: List<PointRow>): String {
        val array = JSONArray()
        points.forEach { array.put(encodeOne(it)) }
        return array.toString()
    }

    fun timestamp(millis: Long): String = iso8601.format(Date(millis))

    fun encodeOne(point: PointRow): JSONObject = JSONObject().apply {
        put("id", point.id)
        put("lat", point.lat)
        put("lon", point.lon)
        put("accuracy", point.accuracy)
        // `put(null)` removes a key, so absent sensors go in as JSONObject.NULL
        // to keep the key present with a null value, as the spec requires.
        put("altitude", point.altitude ?: JSONObject.NULL)
        put("speed", point.speed ?: JSONObject.NULL)
        put("heading", point.heading ?: JSONObject.NULL)
        put("recorded_at", timestamp(point.recordedAtMillis))
        put("is_mock", point.isMock)
        put("battery_level", point.batteryLevel ?: JSONObject.NULL)
    }
}

File: ios/background_geo_tracker/Sources/background_geo_tracker/PointJson.swift
import Foundation

/// Builds the upload body. The wire format is fixed by the design spec.
enum PointJson {
    private static let formatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter
    }()

    static func timestamp(_ millis: Int64) -> String {
        formatter.string(
            from: Date(timeIntervalSince1970: Double(millis) / 1000)
        )
    }

    static func encodeOne(_ row: GeoPointRow) -> [String: Any] {
        // NSNull, not a dropped key: the spec requires the key to be present
        // with a null value when a sensor has nothing to report.
        [
            "id": row.id,
            "session_id": row.sessionId,
            "lat": row.lat,
            "lon": row.lon,
            "accuracy": row.accuracy,
            "altitude": row.altitude ?? NSNull(),
            "speed": row.speed ?? NSNull(),
            "heading": row.heading ?? NSNull(),
            "recorded_at": timestamp(row.recordedAtMillis),
            "is_mock": row.isMock,
            "battery_level": row.batteryLevel ?? NSNull(),
        ]
    }

    static func encode(_ rows: [GeoPointRow]) -> Data {
        let payload = rows.map(encodeOne)
        return (try? JSONSerialization.data(withJSONObject: payload))
            ?? Data("[]".utf8)
    }
}
